File: sorento_crm_frontend/app/(protected)/dealer-kit/services/editionService.ts
/**
 * Catalogue Editions - the approval cycle over a page (S2.5).
 *
 * ---------------------------------------------------------------------------
 * CONTRACT - `app/api/v1/dealer_kit/editions.py`, all under `/api/v1/dealer-kit`.
 * ---------------------------------------------------------------------------
 *
 * GET    /editions?pageId=            -> Edition[], newest first
 * GET    /editions/{id}               -> Edition
 * POST   /editions {pageId, name}     -> 201 Edition, 409 if one is already open
 * POST   /editions/{id}/submit        -> Edition   (page.edit)
 * POST   /editions/{id}/approve       -> Edition   (edition.approve)
 * POST   /editions/{id}/reject {reason} -> Edition (edition.approve)
 * POST   /editions/{id}/reopen        -> Edition   (page.edit)
 * POST   /editions/{id}/publish       -> Edition   (page.publish)
 *
 * **Three rights, and they are deliberately not the same one.** `page.edit`
 * starts and submits, `edition.approve` decides, `page.publish` moves the
 * published label. A Designer therefore gets a 403 on approve INCLUDING on
 * their own Edition (AC-L3), which is the whole point of the workflow.
 *
 * **Approving publishes nothing.** It records that a human read the catalogue
 * and WHICH version they read. Marking it done is what readers see (AC-L7).
 *
 * **`status` is the graph's key, never a status id.** The label travels beside
 * it so no screen keeps its own copy of the vocabulary.
 */

import { apiFetch } from '@/lib/api';
import { extractApiError } from '@/lib/api-client';

const BASE = '/api/v1/dealer-kit/editions';

/** The five states of the seeded `dealer_kit_edition` graph. */
export type EditionStatus =
  | 'draft'
  | 'pending_approval'
  | 'approved'
  | 'rejected'
  | 'done';

export interface Edition {
  id: string;
  pageId: string;
  /** The catalogue's name. Shown instead of the page id, which is a uuid. */
  pageName: string | null;
  name: string;
  status: EditionStatus;
  statusLabel: string;
  approvedVersionId: string | null;
  doneVersionId: string | null;
  previousEditionId: string | null;
  submittedAt: string | null;
  approvedAt: string | null;
  rejectionReason: string | null;
  createdAt: string;
}

function toEdition(wire: Partial<Edition>): Edition {
  return {
    id: String(wire.id ?? ''),
    pageId: String(wire.pageId ?? ''),
    pageName: wire.pageName ?? null,
    name: wire.name ?? 'Untitled edition',
    status: (wire.status as EditionStatus) ?? 'draft',
    statusLabel: wire.statusLabel ?? 'Draft',
    approvedVersionId: wire.approvedVersionId ?? null,
    doneVersionId: wire.doneVersionId ?? null,
    previousEditionId: wire.previousEditionId ?? null,
    submittedAt: wire.submittedAt ?? null,
    approvedAt: wire.approvedAt ?? null,
    rejectionReason: wire.rejectionReason ?? null,
    createdAt: wire.createdAt ?? '',
  };
}

export async function listEditions(pageId?: string): Promise<Edition[]> {
  const path = pageId ? `${BASE}?pageId=${encodeURIComponent(pageId)}` : BASE;
  const response = await apiFetch(path);
  if (!response.ok) {
    throw new Error(await extractApiError(response, 'Could not load editions'));
  }
  const rows = (await response.json()) as Partial<Edition>[];
  return (Array.isArray(rows) ? rows : []).map(toEdition);
}

export async function getEdition(editionId: string): Promise<Edition> {
  const response = await apiFetch(`${BASE}/${editionId}`);
  if (!response.ok) {
    throw new Error(await extractApiError(response, 'Could not load this edition'));
  }
  return toEdition(await response.json());
}

export async function createEdition(input: {
  pageId: string;
  name: string;
  previousEditionId?: string;
}): Promise<Edition> {
  const response = await apiFetch(BASE, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(input),
  });
  if (!response.ok) {
    // The 409 body already says what to do about it ("finish or reject the one
    // in progress"), so it is passed through rather than replaced.
    throw new Error(await extractApiError(response, 'Could not start this edition'));
  }
  return toEdition(await response.json());
}

/** Every transition is a POST to a named action. One helper, one contract. */
async function act(
  editionId: string,
  action: 'submit' | 'approve' | 'reject' | 'reopen' | 'publish',
  body?: Record<string, unknown>,
): Promise<Edition> {
  const response = await apiFetch(`${BASE}/${editionId}/${action}`, {
    method: 'POST',
    ...(body
      ? { headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body) }
      : {}),
  });
  if (!response.ok) {
    throw new Error(await extractApiError(response, `Could not ${action} this edition`));
  }
  return toEdition(await response.json());
}

export const submitEdition = (id: string) => act(id, 'submit');
export const approveEdition = (id: string) => act(id, 'approve');
export const rejectEdition = (id: string, reason: string) =>
  act(id, 'reject', { reason });
export const reopenEdition = (id: string) => act(id, 'reopen');
export const publishEdition = (id: string) => act(id, 'publish');

File: sorento_crm_backend/app/services/dealer_kit/edition_service.py
"""The Edition's transitions (S2.5.2).

Every move goes through ``status_service.assert_transition_allowed`` rather
than an ``if``. The graph seeded by migration 318 is the authority on what is
legal, so adding a state later is a seeding change and not a hunt through this
file for the branch that also needs it.

**``done`` is the only thing that publishes** (AC-L7). Approving does not move
the ``published`` label; it records that a human read the catalogue. Whoever
publishes then decides WHEN, and that is a separate permission because moving
the label is what a reader actually sees.

**Both version ids are recorded, and they are not the same fact.**
``approved_version_id`` is what the Approver read; ``done_version_id`` is what
went live. They are usually equal. Storing only one would leave the deferred
price-drift work (AC-L4 to AC-L6) with nothing to compare - see
``PLAN-edition-approval.md``.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.models.dealer_kit import Edition, PageVersion
from app.models.status import Status
from app.modules.dealer_kit.bootstrap import EDITION_ENTITY
from app.services import status_service
from app.services.dealer_kit import page_service
from app.services.error_handler import AppException

DRAFT = "draft"
PENDING_APPROVAL = "pending_approval"
APPROVED = "approved"
REJECTED = "rejected"
DONE = "done"


def _status_by_key(db: Session, key: str) -> Status:
    """The status row for a key in the DEFAULT graph.

    A missing row is a 500-shaped problem, not a user error: the graph is
    seeded by migration and its absence means the migration did not run. Said
    plainly rather than surfacing as an AttributeError three frames down.
    """
    graph = status_service.resolve_graph(db, EDITION_ENTITY, None)
    status = graph.by_key(key)
    if status is None:
        raise AppException(
            status_code=500,
            message=(
                f"The catalogue Edition status '{key}' is missing. "
                "Its status graph has not been seeded."
            ),
        )
    return status


def get_edition(db: Session, edition_id: str) -> Edition:
    """The Edition, or a 404.

    Company scope is applied by the ORM filter, so another company's Edition is
    a 404 here and never a 403 - a 403 confirms the id exists.
    """
    edition = db.query(Edition).filter(Edition.id == edition_id).first()
    if edition is None:
        raise AppException(status_code=404, message="Edition not found")
    return edition


def list_editions(db: Session, *, page_id: Optional[str] = None) -> list[Edition]:
    query = db.query(Edition)
    if page_id:
        query = query.filter(Edition.page_id == page_id)
    return query.order_by(Edition.created_at.desc()).all()


def create_edition(
    db: Session,
    *,
    page_id: str,
    name: str,
    previous_edition_id: Optional[str] = None,
    user_id: Optional[str] = None,
) -> Edition:
    """Start a revision cycle over a page, at ``draft``.

    The one-open-per-page rule is a partial unique index, not a check here: a
    service-level "is one already open" read races itself between two requests
    and the loser writes the second open Edition anyway. So the write is
    attempted and the database's answer is translated - which also means the
    guarantee holds for anything else that ever inserts a row.
    """
    page_service.get_page(db, page_id)  # 404s if it is not this company's page

    initial = _status_by_key(db, DRAFT)
    edition = Edition(
        page_id=page_id,
        name=name,
        status_id=initial.id,
        status_key=initial.key,
        previous_edition_id=previous_edition_id,
        created_by=user_id,
    )
    db.add(edition)
    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        raise AppException(
            status_code=409,
            message=(
                "This catalogue already has an Edition in progress. "
                "Finish or reject it before starting another."
            ),
        )
    db.refresh(edition)
    return edition


def _move(
    db: Session,
    edition: Edition,
    to_key: str,
    *,
    apply: Optional[callable] = None,
) -> Edition:
    """Move an Edition to ``to_key``, or refuse because the graph says so.

    ``apply`` stamps whatever else the transition records, and runs only AFTER
    the move is known to be legal - so a rejected transition cannot leave an
    ``approved_by`` behind on a record that never moved.
    """
    target = _status_by_key(db, to_key)
    status_service.assert_transition_allowed(
        db,
        EDITION_ENTITY,
        from_status_id=edition.status_id,
        to_status_id=target.id,
    )

    edition.status_id = target.id
    # Carried in lockstep. The partial unique index reads the KEY, so a move
    # that updated only the id would leave a done Edition still occupying its
    # page's one open slot.
    edition.status_key = target.key
    if apply is not None:
        apply()

    try:
        db.commit()
    except IntegrityError:
        db.rollback()
        raise AppException(
            status_code=409,
            message=(
                "This catalogue already has an Edition in progress. "
                "Finish or reject it before reopening this one."
            ),
        )
    db.refresh(edition)
    return edition


def submit(db: Session, edition_id: str, *, user_id: Optional[str] = None) -> Edition:
    """Designer: this is ready for somebody to look at."""
    edition = get_edition(db, edition_id)

    def _stamp() -> None:
        edition.submitted_by = user_id
        edition.submitted_at = datetime.utcnow()
        # Cleared on re-submission: a reason from the previous round shown
        # beside a fresh submission reads as a fresh rejection.
        edition.rejection_reason = None

    return _move(db, edition, PENDING_APPROVAL, apply=_stamp)


def approve(db: Session, edition_id: str, *, user_id: Optional[str] = None) -> Edition:
    """Approver: I have read this.

    Records WHICH version was read. The Approver is signing off a document, not
    a page id, and the page's newest version is what the review screen showed
    them - so that is what is stamped, and a later edit is a different document
    that has to come back round.
    """
    edition = get_edition(db, edition_id)
    latest = _latest_version(db, edition.page_id)

    def _stamp() -> None:
        edition.approved_by = user_id
        edition.approved_at = datetime.utcnow()
        edition.approved_version_id = latest.id if latest else None
        edition.rejection_reason = None

    return _move(db, edition, APPROVED, apply=_stamp)


def reject(
    db: Session, edition_id: str, *, reason: str, user_id: Optional[str] = None
) -> Edition:
    """Approver: no, and here is why.

    The reason is required by the route's schema and re-checked here. A
    rejection with no reason is a rejection the Designer cannot act on, and
    this is the last place that can refuse one.
    """
    reason = (reason or "").strip()
    if not reason:
        raise AppException(
            status_code=422,
            message="Say why it is being rejected. The Designer sees this.",
        )

    edition = get_edition(db, edition_id)

    def _stamp() -> None:
        edition.rejection_reason = reason
        edition.approved_by = None
        edition.approved_at = None

    return _move(db, edition, REJECTED, apply=_stamp)


def reopen(db: Session, edition_id: str, *, user_id: Optional[str] = None) -> Edition:
    """Designer picking rejected work back up.

    The reason is deliberately NOT cleared here - it is cleared on the next
    submission. A Designer editing a rejected Edition should still be able to
    see what they were told.
    """
    return _move(db, get_edition(db, edition_id), DRAFT)


def publish(db: Session, edition_id: str, *, user_id: Optional[str] = None) -> Edition:
    """Move the published label, and finish the Edition. AC-L7.

    The ONLY transition that publishes anything. The label move happens before
    the status move commits, so an Edition can never read ``done`` while the
    label still points somewhere else - the failure that would have a catalogue
    marked live that no reader can see.
    """
    edition = get_edition(db, edition_id)
    latest = _latest_version(db, edition.page_id)
    if latest is None:
        raise AppException(
            status_code=422,
            message=(
                "This catalogue has no saved version to publish. "
                "Save the page first."
            ),
        )

    def _stamp() -> None:
        edition.done_version_id = latest.id
        page_service.move_label(
            db,
            edition.page_id,
            page_service.PUBLISHED,
            version_id=latest.id,
            user_id=user_id,
        )

    return _move(db, edition, DONE, apply=_stamp)


def _latest_version(db: Session, page_id: str) -> Optional[PageVersion]:
    return (
        db.query(PageVersion)
        .filter(PageVersion.page_id == page_id)
        .order_by(PageVersion.version.desc())
        .first()
    )
